[pidcontroller] Initial Contribution (#9512)
openhab/openhab-addons: bundles/org.openhab.automation.pidcontroller/README.md

| Name         | Type | Description                                                          | Required |
|--------------|------|----------------------------------------------------------------------|----------|
| `output`     | Item | Name of the output Item (e.g. the valve actuator 0-100%)             | Y        |
| `pInspector` | Item | Name of the debug Item for the current P part                        | N        |
| `iInspector` | Item | Name of the debug Item for the current I part                        | N        |
| `dInspector` | Item | Name of the debug Item for the current D part                        | N        |
| `eInspector` | Item | Name of the debug Item for the current regulation difference (error) | N        |

You can view the internal P, I and D parts of the controller with the inspector Items.
These values are useful when tuning the controller.
They are updated everytime the output is updated.

## Proportional (P) Gain Parameter

Parameter: `kp`

A value of 0 disables the P part.

A value of 1 sets the output to the current setpoint deviation (error).
E.g. the setpoint is 25°C and the measured value is 20°C, the output will be set to 5.

> If the output is the opening of a valve in %, you might want to set this parameter to higher values (`Kp=10` would result in 50%).

Reviewer: ```suggestion
If the output is the opening of a valve in %, you might want to set this parameter to higher values (`kp=10` would result in 50%).
```

This makes the regulation more resilient against oscillations, i.e. bigger `kp` and `ki` values can be set.

A value of 0 disables the D part.

A value of 1 sets the output to the difference between the last setpoint deviation (error) and the current.
E.g. the setpoint is 25°C and the measured value is 20°C (error=5°C).
When the temperature drops to 10°C due to an opened window (error=15°C), the output is set to 15°C - 5°C = 10.

## Derivative Time Constant (D-T1) Parameter

Parameter: `kdTimeConstant`

The purpose of this parameter is to slow down the impact of the D part.

This parameter behaves like a [low-pass](https://en.wikipedia.org/wiki/Low-pass_filter) filter.
The D part will become 63% of its actual value after `kdTimeConstant` seconds and 99% after 5 times `kdTimeConstant`. E.g. `kdTimeConstant` is set to 10s, the D part will become 99% after 50s.

Higher values lead to a longer lasting impact of the D part (stretching) after a change in the setpoint deviation (error).
The "stretching" also results in a lower amplitude, i.e. if you increase this value, you might want to also increase `kd` to keep the height of the D part at the same level.

## Tuning

> Tuning the `Kp`, `Ki` and `Kd` parameters can be done by appling science.

Reviewer: ```suggestion
Tuning the `Kp`, `Ki` and `Kd` parameters can be done by applying science.
```

It can also be done by heuristic methods like the [Ziegler–Nichols method](https://en.wikipedia.org/wiki/Ziegler%E2%80%93Nichols_method).
But it can also be done by trial and error.
This results in quite reasonable working systems in most cases.
So, this will be described in the following.

> To be able to proceed with this method, you need to visualization the input and the output value of the PID controller over time.

Reviewer: ```suggestion
To be able to proceed with this method, you need to visualize the input and the output value of the PID controller over time.
```

It's also good to visualize the individual P, I and D parts (these are forming the output value) via the inspector Items.
The visualization can be done by the analyze function in Main UI or by adding a persistence and use Grafana for example.

After you added a [Rule](https://www.openhab.org/docs/configuration/rules-dsl.html) with above trigger and action module and configured those, proceed with the following steps:

> *Notice:* A good starting point for the derivative time constant `kdTimeConstant` is the response time of the control loop.
E.g. the time it takes from opening the heater valve and seeing an effect of the measured temperature.

1. Set `kp`, `ki` and `kd` to 0

> 2. Increase `kp` as long as the system starts to oscillate (continuous over- and undershoot)

Reviewer: ```suggestion
2. Increase `kp` until the system starts to oscillate (continuous over- and undershoot)
```
